content/2022-04-29-firmware-updates-part-1/index.md

```diff
@@ -0,0 +1,160 @@
+
+The bootloader flash memory is divided into logical areas named `Partitions`:
+
+<figure>
+    <img src="partitions.png" alt="Bootloader partitions" />
+    <figcaption>Bootloader partitions</figcaption>
+</figure>
+
+Each partition is an area of flash on a microcontroller, but only the `Bootloader` and `Active` partitions need to be on the same physical flash storage. This means you can use off-chip flash to store the firmware update if desired.
+
+The `Bootloader` partition contains the bootloader application itself, usually around 8kB if disabling debug logging. The `Bootloader State` partition contains a flag indicating if firmware within the `Active` and `DFU` partitions should be swapped or not, as well as a progress index used by the swap and revert algorithm.
+
+The `Active` partition contains the currently running firmware, and the bootloader will always jump to the application at the start of the `Active`  partition. The `DFU` partition contains the "next" firmware to be updated, until the update is signaled, at which point the bootloader will gradually swap it with the "old" firmware.
+
+## Initial boot
+
+At initial boot, the bootloader performs the following operations:
+
+1. Read current bootloader state magic word.
+1. If no "SWAP" magic bytes found, boot the current firmware.
+1. If "SWAP" magic bytes found, perform the swap between the Active and DFU firmwares.
+1. Boot the new current firmware.
+
+## Updating the firmware
+
+An application that wants to be firmware update capable will need to create an instance of a `FirmwareUpdater`. This is provided by the `embassy-boot` library, with defaults available if using the `embassy-boot-nrf` or `embassy-boot-stm32` libraries. The `FirmwareUpdater` is capable of the following operations:
+
+* Write firmware to any offset within the `DFU` partition.
+* Mark the current running firmware as 'OK', to prevent a firmware rollback to occur.
+* Mark the current firwmare to be swapped by the new firmware.
+
+How the firmware gets to the device is not the responsibility of the `FirmwareUpdater`, which leaves this problem to the application itself. In the next blog post, we will cover different ways you can get the firmware to the device.
+
+Once the new firmware is written, and marked to be swapped with the current firmware, the application needs to reset the device. Upon reset, the device will enter the bootloader once again, and attempt to swap the firmware before booting the next version.
+
+NOTE: By structuring your application accordingly, you may update the firmware in parallel while running other tasks. This reduces the downtime of your application to a minimum.
+
+## Swap algorithm
+
+At the core of the bootloader is the bank swapping algorithm, which is not tied to any specific platform. The algorithm keeps an internal state of the copy progress in flash, using 1 word per page to spread the writes.
```

Comment:
Rather than erase + write the progress to only 1-2 locations in flash, it uses a separate location for each index element ("spreading"), which requires only erasing before each swap, rather than each index counter update. I'll elaborate it a bit.